python/cudf_polars/cudf_polars/experimental/sort.py

```diff
@@ -4,19 +4,350 @@
 
 from __future__ import annotations
 
-from typing import TYPE_CHECKING
+from typing import TYPE_CHECKING, Any, TypedDict
 
-from cudf_polars.dsl.ir import Sort
-from cudf_polars.experimental.base import PartitionInfo
-from cudf_polars.experimental.dispatch import lower_ir_node
+import pyarrow as pa
+
+import pylibcudf as plc
+import rmm.mr
+from rmm.pylibrmm.stream import DEFAULT_STREAM
+
+from cudf_polars.containers import DataFrame
+from cudf_polars.dsl.expr import Col
+from cudf_polars.dsl.ir import IR, Sort
+from cudf_polars.dsl.traversal import traversal
+from cudf_polars.dsl.utils.naming import unique_names
+from cudf_polars.experimental.base import PartitionInfo, get_key_name
+from cudf_polars.experimental.dispatch import generate_ir_tasks, lower_ir_node
 from cudf_polars.experimental.repartition import Repartition
-from cudf_polars.experimental.utils import _lower_ir_fallback
+from cudf_polars.experimental.shuffle import _simple_shuffle_graph
+from cudf_polars.experimental.utils import _concat, _lower_ir_fallback
 
 if TYPE_CHECKING:
-    from collections.abc import MutableMapping
+    from collections.abc import MutableMapping, Sequence
 
-    from cudf_polars.dsl.ir import IR
+    from cudf_polars.dsl.expr import NamedExpr
     from cudf_polars.experimental.dispatch import LowerIRTransformer
+    from cudf_polars.typing import Schema
+    from cudf_polars.utils.config import ConfigOptions
+
+
+def find_sort_splits(
+    tbl: plc.Table,
+    split_candidates_unordered: plc.Table,
+    my_part_id: int,
+    num_partitions: int,
+    column_order: Sequence[plc.types.Order],
+    null_order: Sequence[plc.types.NullOrder],
+) -> list[int]:
+    """
+    Find the local sort splits given all (global) split candidates.
+
+    The reason for much of the complexity is to get the result sizes as
+    precise as possible even when e.g. all values are equal.
+    In other words, this goes through extra effort to split the data at the
+    precise boundaries (which includes part_id and local_row_number).
```

Comment:
Okay - I was initially confused by the fact that this function wasn't already applied to the concatenated result in `_sort_boundaries_graph` (before being broadcasted to the shuffler insertion).

If I understand correctly, this is the motivation. We may be doing redundant work for every input partition, **but** this allows us to handle pathological data distributions. How much of an overhead do you think this is for uniformly-distributed data (where local adjustments are unnecessary)?

Reply:
`_sort_boundaries_graph` (I'll rename to `_split_boundary_candidates_graph`, I think) are the values at which we wish to split.

To find the indices at which to split, we also need the local sorted data, so that is a clear second step (and the one that adds the complexity, so it is not avoidable).

Let me pull out the sort+extract part and move it into `_sort_boundaries_graph`.  I suppose, right now dask is likely pulling it all into one worker anyway and then distributing, so might as well sort+extract there.

Plus, I thought a bit more about a case where we have one partition per worker/GPU where repeating that work locally seems fine, but if we have many partitions per worker/GPU it may be nice not do.